first time commit (#999)
DataDog/integrations-extras: tyk/README.md

# Agent Check: tyk

## Overview

![tyk](https://tyk.io/wp-content/uploads/2021/07/tyk_logo_no_bg.png)

![tyk-in-action](./images/tyk-in-lab.png?raw=true "Tyk's Lab")

> This check collects metrics from [Tyk-gateway][https://github.com/TykTechnologies/tyk].

Reviewer: All links should be numbered notation, for our translation pipelines to work properly. It looks like you do have some at the end of the file, please update the rest of them to follow that format.

# Agent Check: tyk

## Overview

![tyk][3]

![tyk-in-action][5]

Datadog can collect and display errors response time, duration, latency and monitor performance of the api traffic
in [tyk][1] to easily discover issues in your APIs or your consumers.

Tyk has a built-in Datadog integration that collects metrics from [Tyk API gateway][7].
Tyk API gateway records for you all the traffic that it's processing and can send it for you to Datadog and build 
dashboards around it.

> ### How does it work behind the scenes

Reviewer: ```suggestion
### How it works
```

### Configuration

#### Setup *Tyk-pump*: 
To set a Datadog pump follow the instructions in this [link][18]
Here is an example of Datadog pump configuration in `pump.conf`. 

``` json
pump.conf:
...
   "dogstatsd": {
      "type": "dogstatsd",
      "meta": {
        "address": "dd-agent:8126",
        "namespace": "tyk",
        "async_uds": true,
        "async_uds_write_timeout_seconds": 2,
        "buffered": true,
        "buffered_max_messages": 32,
        "sample_rate": 0.9999999999,
        "tags": [
          "method",
          "response_code",
          "api_version",
          "api_name",
          "api_id",
          "org_id",
          "tracked",
          "path",
          "oauth_id"
        ]
      }
    },
```
This [example][19] was taken from [Tyk-demo][20] project, an open source project that spins up a full tyk platform in
one command and offers ready-made examples, among which is the Datadog example.
To run this integration do `up.sh analytics-datadog`

#### **Setup Datadog Agent**: 
Tyk's integration makes use of StatsdogD. `Tyk-pump` writes over UDP to StatsdogD that is part of the agent.

> Please set up the following Datadog environment variables in your environment:

Author: `Tyk-pump` uses UDP to tell DogstatsD that it is part of the agent` - this has a different meaning that is not right in this context

| DD_DOGSTATSD_SOCKET | /var/run/docker.sock | Path to the Unix socket to listen to.  Docker compose mounts this path |
| DD_DOGSTATSD_ORIGIN_DETECTION | true | Enable container detection and tagging for unix socket metrics |
| DD_DOGSTATSD_NON_LOCAL_TRAFFIC | true | Listen to DogStatsD packets from other containers (required to send custom metrics) |
| DD_AGENT_HOST | dd-agent | Name of the agent host in docker |
| DD_AC_EXCLUDE | redis | To exclude DD redis checks |
| DD_CONTAINER_EXCLUDE | true | To exclude docker checks for the DD agent|

This is the [doc][21] to set up the agent.

[Restart the Agent][4] after setting the environment variables

### Validation

Create a dashboard (or import our sample[]) and add a widget. In the section `Graph your data` under the `metric`
option, start typing the namespace you chose for the pump in the config `pump.conf` under `dogstatsd.namespace`. 
In the example above it was `tyk`. Once you start typing you will see all the available metrics - `tyk.request_time.avg` etc.

## Data Collected

### Metrics

See [metadata.csv][6] for a list of metrics provided by this check.

### Dashboards

> With Datadog, you can easily create nice dashboards that displays statistics about your API services and

Reviewer: ```suggestion
With Datadog, you can create dashboards that display statistics about your API services and
```

their consumption.

Here's an example for such a dashboard:
![Tyk Analytics dashboard example][10]

**Note: You can [import][11] the above dashboard and use it as an example or baseline for your own dashboard.**

### Events

The tyk integration does not include any events ATM but you can add them on your own.

### Service Checks

The tyk integration does not include any service checks.

## Troubleshooting

Need help? Contact [Datadog support][8].

[1]: https://tyk.io/
[2]: https://docs.datadoghq.com/agent/kubernetes/integrations/
[3]: https://tyk.io/wp-content/uploads/2021/07/tyk_logo_no_bg.png
[4]: https://docs.datadoghq.com/agent/guide/agent-commands/#start-stop-and-restart-the-agent
[5]: ./images/tyk-in-lab.png?raw=true "Tyk's Lab"

> [6]: ./metadata.csv

Reviewer: ```suggestion
[6]: https://github.com/DataDog/integrations-extras/blob/master/tyk/metadata.csv
```

[7]: https://github.com/TykTechnologies/tyk
[8]: https://docs.datadoghq.com/help/
[9]: https://docs.datadoghq.com/developers/dogstatsd/?tab=hostagent#pagetitle
[10]: ./images/datadog-tyk-analytics-dashboard.jpg

> [11]: ./assets/dashboards/tyk_analytics_canvas.json

Reviewer: ```suggestion
[11]: https://github.com/DataDog/integrations-extras/blob/master/tyk/assets/dashboards/tyk_analytics_canvas.json
```

**metric** option, start typing the namespace you chose for the pump in the config `pump.conf` under `dogstatsd.namespace`.
In the example above it was `tyk`. Once you start typing you will see all the available metrics - `tyk.request_time.avg` etc.

## Data Collected

### Metrics

See [metadata.csv][6] for a list of metrics provided by this check.

### Dashboards

With Datadog, you can create dashboards that display statistics about your API services and their consumption.

Here's an example for such a dashboard:
![Tyk Analytics dashboard example][10]

**Note: You can [import][11] the above dashboard and use it as an example or baseline for your own dashboard.**

### Events

The tyk integration does not include any events.

### Service Checks

The tyk integration does not include any service checks.

## Troubleshooting

Need help? Contact [Datadog support][8].

[1]: https://tyk.io/
[2]: https://docs.datadoghq.com/agent/kubernetes/integrations/
[3]: https://tyk.io/wp-content/uploads/2021/07/tyk_logo_no_bg.png
[4]: https://docs.datadoghq.com/agent/guide/agent-commands/#start-stop-and-restart-the-agent

> [5]: ./images/tyk-in-lab.png?raw=true "Tyk's Lab"

Reviewer: ```suggestion
[5]: https://raw.githubusercontent.com/DataDog/integrations-extras/master/tyk/images/tyk-in-lab.png
```

> [5]: https://raw.githubusercontent.com/DataDog/integrations-extras/master/tyk/images/tyk-in-lab.png

Reviewer 2: This link is broken/not in the `images` directory.

Author: @fanny-jiang I'm not sure how to fix this. If I try my fork it works `https://raw.githubusercontent.com/letzya/integrations-extras/master/tyk/images/tyk-in-lab.png`

Author: Can you please help me understand which one isn't in the images directory?
